### PR TITLE
DAOS-2531 utils: new daos tool for pool, container, object mgmt

### PR DESCRIPTION
**Pull request by kccain**
This change introduces a unified user administration tool 'daos'.
The preliminary support provided in this change includes read-only
pool management features, and support for container create, destroy.
That is, it includes some of the features of the 'dmg' tool and
all features of the 'dcont' tool. 'dcont' will be removed in a
subsequent change, effectively replaced by 'daos'. Help/usage
output is provided for the features implemented now, and for pool
and container commands to be added over time in future development.

Some command-line arguments needed to support unified namespace
(path, type, oclass, chunk_size) are included in this patch.
However, full consolidation of the existing 'duns' tool features
into 'daos' will occur in a subsequent change.

Signed-off-by: Ken Cain <kenneth.c.cain@intel.com>

### CODE REVIEW

File: src/utils/daos.c [hunk 1]
```diff
@@ -0,0 +1,870 @@
...
+static int
+help_hdlr(struct cmd_args_s *ap)
+{
+	FILE *stream;
+
+	assert(ap != NULL);
+
+	stream = (ap->ostream != NULL) ? ap->ostream : stdout;
+
+	fprintf(stream, "\
```
[the remark this answers is not in the document]
Comment (kccain) [reply src/utils/daos.c]: following model found in dmg.c and dcont.c

File: src/utils/daos.c [hunk 2]
```diff
@@ -0,0 +1,870 @@
...
+	fprintf(stream, "\n\
+pool commands:\n\
+	  list-containers  list all containers in pool\n\
+	  list-cont\n\
+	  query            query a pool\n\
+	  stat             get pool statistics\n\
+	  list-attrs       list pool user-defined attributes\n\
+	  get-attr         get pool user-defined attribute\n");
+
+	fprintf(stream, "\
```
[the remark this answers is not in the document]
Comment (kccain) [reply src/utils/daos.c]: following model found in dmg.c and dcont.c

File: src/utils/daos.c [hunk 3]
```diff
@@ -0,0 +1,870 @@
...
+	  get-attr         get container user-defined attribute\n\
+	  set-attr         set container user-defined attribute\n\
+	  create-snap      create container snapshot (optional name)\n\
+			   at most recent committed epoch\n\
+	  list-snaps       list container snapshots taken\n\
+	  destroy-snap     destroy container snapshots\n\
+			   by name, epoch or range\n\
+	  rollback         roll back container to specified snapshot\n");
+
+	fprintf(stream, "\
```
[the remark this answers is not in the document]
Comment (kccain) [reply src/utils/daos.c]: following model found in dmg.c and dcont.c

File: src/utils/daos.c [hunk 4]
```diff
@@ -0,0 +1,870 @@
...
+pool options:\n\
+	--pool=UUID        pool UUID \n\
+	--group=STR        pool server process group (\"%s\")\n\
+	--svc=RANKS        pool service replicas like 1,2,3\n\
+	--attr=NAME        pool attribute name to get\n",
+	default_group);
+
+	fprintf(stream, "\n\
```
[the remark this answers is not in the document]
Comment (kccain) [reply src/utils/daos.c]: following model found in dmg.c and dcont.c

File: src/utils/daos.c [hunk 5]
```diff
@@ -0,0 +1,870 @@
...
+	stream = (ap->ostream != NULL) ? ap->ostream : stdout;
+
+	fprintf(stream, "\
+usage: daos RESOURCE COMMAND [OPTIONS]\n\
+resources:\n\
+	  pool             pool\n\
+	  container (cont) container\n\
+	  help             print this message and exit\n");
+
+	fprintf(stream, "\n\
```
[the remark this answers is not in the document]
Comment (kccain) [reply src/utils/daos.c]: following model found in dmg.c and dcont.c

**Comment by kccain** [comment]
This PR is expected to resolve the problem with  build/Build RPM on CentOS 7:
https://github.com/daos-stack/daos/pull/558.

File: src/utils/daos.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (kccain) [reply src/utils/daos.c]: following model found in dmg.c and dcont.c

File: src/utils/daos.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (kccain) [reply src/utils/daos.c]: following model found in dmg.c and dcont.c

File: src/utils/daos.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (kccain) [reply src/utils/daos.c]: following model found in dmg.c and dcont.c

File: src/utils/daos.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (kccain) [reply src/utils/daos.c]: following model found in dmg.c and dcont.c

File: src/utils/daos.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (kccain) [reply src/utils/daos.c]: following model found in dmg.c and dcont.c

File: src/utils/daos.c [hunk 6]
```diff
@@ -0,0 +1,870 @@
...
+	char		*s, *s_saved;
+	char		*end_s;
+	char		*p;
+
+	s = s_saved = strdup(ap->epcrange_str);
+	if (s == NULL) {
+		fprintf(stderr, "strdup() failed\n");
+		goto out;
+	}
+
```
Comment (mchaarawi) [inline src/utils/daos.c]: can't you just use sscanf()?

File: src/utils/daos.c [hunk 7]
```diff
@@ -0,0 +1,870 @@
...
+	return size;
+}
+
+/* acopy_str()
+ * Allocate dest string and copy source contents.
+ * Return NULL if not possible. Caller ensures src is not NULL.
+ */
+static char *
+acopy_str(const char *src)
```
Comment (mchaarawi) [inline src/utils/daos.c]: sorry, maybe i miss something, but why do you need this if you have strdup()?

File: src/utils/daos.c [hunk 8]
```diff
@@ -0,0 +1,870 @@
...
+	--eprange=B-E      container epoch range (destroy-snap)\n");
+	return 0;
+}
+
+int
+main(int argc, char *argv[])
+{
+	int			rc = 0;
+	command_hdlr_t		hdlr = NULL;
+	struct cmd_args_s	dargs;
```
Comment (mchaarawi) [inline src/utils/daos.c]: dargs = {0};
then you don't need the init function?

File: src/utils/daos.c [hunk 9]
```diff
@@ -0,0 +1,870 @@
...
+		struct daos_pool_space		*ps = &pinfo.pi_space;
+		struct daos_rebuild_status	*rstat = &pinfo.pi_rebuild_st;
+		int				 i;
+
+		rc = daos_pool_connect(ap->pool_uuid, ap->group,
+				       ap->mdsrv, DAOS_PC_RO, &pool,
+				       NULL /* info */, NULL /* ev */);
+		if (rc != 0) {
+			fprintf(stderr, "failed to connect to pool: %d\n", rc);
+			goto bad_pool_connect;
```
Comment (mchaarawi) [inline src/utils/daos.c]: use D_GOTO

File: src/utils/daos.c [hunk 10]
```diff
@@ -0,0 +1,870 @@
...
+
+	/* TODO implement the following:
+	 * op == POOL_LIST_CONTAINERS
+	 * op == POOL_STAT
+	 * op == POOL_GET_PROP
+	 * op == POOL_GET_ATTR
+	 * op == POOL_LIST_ATTRS
+	 */
+
+bad_pool_query:
```
Comment (mchaarawi) [inline src/utils/daos.c]: i wouldn't call this "bad", since even if everything succeeds above, you still get here.

File: src/utils/daos.c [hunk 11]
```diff
@@ -0,0 +1,870 @@
...
+
+	/* TODO: decide if for container operations the code should
+	 * check that a container UUID or path has been provided,
+	 * and if a path, that container UUID can be looked up (UNS).
+	 * Or, if that checking should be in container command handling.
+	 */
+	return 0;
+
+bad_opt_free_mdsrv:
+	daos_rank_list_free(ap->mdsrv);
```
Comment (mchaarawi) [inline src/utils/daos.c]: you have to free this even in success condition..

File: src/utils/daos.c [hunk 12]
```diff
@@ -0,0 +1,870 @@
...
+		rc = daos_cont_create(pool, ap->cont_uuid, NULL, NULL);
+		if (rc != 0) {
+			fprintf(stderr, "failed to create container: %d\n", rc);
+			goto bad_cont_create;
+		}
+		fprintf(stdout, "Successfully created container "DF_UUIDF"\n",
+			DP_UUID(ap->cont_uuid));
+	}
+
+	if (op != CONT_DESTROY) {
```
Comment (mchaarawi) [inline src/utils/daos.c]: the next 2 if statements can be merged

File: src/utils/daos.c [hunk 13]
```diff
@@ -0,0 +1,870 @@
...
+		rc = daos_cont_open(pool, ap->cont_uuid, DAOS_COO_RW, &coh,
```
Comment (mchaarawi) [inline src/utils/daos.c]: why do you need to open/close the container in case of CREATE op?

File: src/utils/daos.c [hunk 14]
```diff
@@ -0,0 +1,870 @@
...
+	 * op == CONT_DEL_ATTR
+	 * op == CONT_GET_ATTR
+	 * op == CONT_SET_ATTR
+	 * op == CONT_CREATE_SNAP
+	 * op == CONT_LIST_SNAPS
+	 * op == CONT_DESTROY_SNAP
+	 * op == CONT_ROLLBACK
+	 */
+
+bad_cont_destroy:
```
Comment (mchaarawi) [inline src/utils/daos.c]: don't use bad since this is the success path too.

File: src/utils/daos.c [hunk 15]
```diff
@@ -0,0 +1,870 @@
...
+bad_cont_close:
+bad_cont_open:
+bad_cont_create:
+	/* Pool disconnect  in normal and error flows: preserve rc */
+	rc2 = daos_pool_disconnect(pool, NULL);
+	if (rc2 != 0) {
```
Comment (mchaarawi) [inline src/utils/daos.c]: please be consistent with single line brackets. either use them, or don't

File: src/utils/daos.c (hunk 9 above)
Comment (kccain) [reply src/utils/daos.c]: Will do - is this a general comment for every goto in the program or only certain cases ?

File: src/utils/daos.c (hunk 8 above)
Comment (kccain) [reply src/utils/daos.c]: need to initialize a few fields to something other than 0 or NULL. Will remove the init function and just put those small number of lines in common_op_parse_hdlr()

File: src/utils/daos.c (hunk 15 above)
Comment (kccain) [reply src/utils/daos.c]: agree, will fix in next patch

File: src/utils/daos.c (hunk 13 above)
Comment (kccain) [reply src/utils/daos.c]: copy/paste from dcont. Will change the conditional to check for op != CREATE && op != DESTROY

File: src/utils/daos.c (hunk 11 above)
Comment (kccain) [reply src/utils/daos.c]: agree, will fix in next patch (won't have bad_ in the label name)

File: src/utils/daos.c (hunk 10 above)
Comment (kccain) [reply src/utils/daos.c]: agree, will fix in next patch

File: src/utils/daos.c (hunk 7 above)
Comment (kccain) [reply src/utils/daos.c]: will replace with strdup calls instead in next patch

File: src/utils/daos.c (hunk 6 above)
Comment (kccain) [reply src/utils/daos.c]: will attempt sscanf() this in next patch. malformed input error catching issues were considered. This also was created starting from some other strtok-based parsing function, daos_rank_list_parse()

File: src/utils/daos.c [hunk 16]
```diff
@@ -0,0 +1,870 @@
...
+
+	/* argv[1] is RESOURCE or "help";
+	 * argv[2] if provided is a resource-specific command
+	 */
+	if (argc <= 2 || strcmp(argv[1], "help") == 0)
+		hdlr = help_hdlr;
+	else if (strcmp(argv[1], "container") == 0)
+		hdlr = cont_op_hdlr;
+	else if (strcmp(argv[1], "cont") == 0)
+		hdlr = cont_op_hdlr;
```
[the remark this answers is not in the document]
Comment (kccain) [reply src/utils/daos.c]: agree, will fix in next patch

File: src/utils/daos.c [hunk 17]
```diff
@@ -0,0 +1,870 @@
...
+	else if (strcmp(argv[1], "pool") == 0)
+		hdlr = pool_op_hdlr;
+
+	if (hdlr == NULL) {
+		dargs.ostream = stderr;
```
[the remark this answers is not in the document]
Comment (kccain) [reply src/utils/daos.c]: I had actually started out with this in mind. But it got a little messy in the sense that help_hdlr() has the same function pointer typedef as the other actual pool, container handler functions.

File: src/utils/daos.c [hunk 18]
```diff
@@ -392,81 +326,90 @@ common_op_parse_hdlr(int argc, char *argv[], struct cmd_args_s *ap)
...
+			ap->group = strdup(optarg);
 			break;
 		case 'p':
 			if (uuid_parse(optarg, ap->pool_uuid) != 0) {
 				fprintf(stderr,
 					"failed to parse pool UUID: %s\n",
 					optarg);
-				return RC_NO_HELP;
+				rc = RC_NO_HELP;
+				D_GOTO(out_free, rc);
```
Comment (mchaarawi) [inline src/utils/daos.c]: you can collapse this (and all other subsequent D_GOTO) into one line:
D_GOTO(out_free, rc = RC_NO_HELP);

File: src/utils/daos.c [hunk 19]
```diff
@@ -0,0 +1,813 @@
...
+			return RC_PRINT_HELP;
+		}
+	} else {
+		/* main() may catch error. Keep this code just in case. */
+		fprintf(stderr, "resource (%s): must be "
+				 "pool or container\n", argv[1]);
+		fflush(stderr);
+		return RC_PRINT_HELP;
+	}
+	cmdname = strdup(argv[2]);
```
Comment (okulkarni) [inline src/utils/daos.c]: Probably a good idea to get all memory management stuff to be done through "gurt" utility methods. Other than malloc and free, "gurt" (cart/src/include/gurt/common.h) also provides wrappers for strdup/strndup (D_STRNDUP) and asprintf (D_ASPRINTF) and realpath (D_REALPATH).

**Pull request merged**